Sources/NIO/BaseSocketChannel.swift

```diff
@@ -469,32 +475,83 @@ class BaseSocketChannel<SocketType: BaseSocketProtocol>: SelectableChannel, Chan
+                assert(!self.isWritePending())
+                ()
+            }
+        }
+    }
+
     /// Flush data to the underlying socket and return if this socket needs to be registered for write notifications.
     ///
-    /// - returns: If this socket should be registered for write notifications. Ie. `IONotificationState.register` if _not_ all data could be written, so notifications are necessary; and `IONotificationState.unregister` if everything was written and we don't need to be notified about writability at the moment.
+    /// - warning: You must not call `flushNow` re-entrantly. Use `flushNowAndRegisterIfNecessary` to do what it says
+    ///            on the tin and also protect against re-entrancy.
+    ///
+    /// - returns: If this socket should be registered for write notifications. Ie. `IONotificationState.register` if
+    ///            _not_ all data could be written, so notifications are necessary; and `IONotificationState.unregister`
+    ///             if everything was written and we don't need to be notified about writability at the moment.
     func flushNow() -> IONotificationState {
         self.eventLoop.assertInEventLoop()
-        // Guard against re-entry as data that will be put into `pendingWrites` will just be picked up by
-        // `writeToSocket`.
-        guard !self.inFlushNow && self.isOpen else {
-            return .unregister
-        }
+        assert(self.isOpen)
+        assert(self.lifecycleManager.isActive)
+
+        // We should not be re-entered anymore because we're only called from `writable()` (which doesn't get called
+        // re-entrantly and `flushNowAndRegisterIfNecessary()` which protects against re-entrancy.
+        precondition(self.flushNowState == .notInFlushNow, "illegal flushNowState \(self.flushNowState)")
 
         do {
             var writeResult: OverallWriteResult? = nil
 
             defer {
                 if writeResult?.writabilityChange == .some(true) {
+                    assert(self.flushNowState == .notInFlushNow)
+                    self.flushNowState = .inFlushNowCallingOut
+                    defer {
+                        assert(self.flushNowState == .inFlushNowCallingOut)
+                        self.flushNowState = .notInFlushNow
+                    }
+
                     // We went from not writable to writable.
                     self.pipeline.fireChannelWritabilityChanged0()
                 }
             }
 
             do {
-                assert(!self.inFlushNow)
-                self.inFlushNow = true
+                assert(self.flushNowState == .notInFlushNow)
+                self.flushNowState = .inFlushNowWriting
                 defer {
-                    self.inFlushNow = false
+                    assert(self.flushNowState == .inFlushNowWriting)
+                    self.flushNowState = .notInFlushNow
                 }
```

Comment:
There are way too many defer blocks floating around now, I think this code needs to be refactored. It's extremely hard to follow the ordering of operations in this relatively straightforward function. I think the call flow is straightforward enough that we can rewrite this with only a single defer block (specifically, the one that sets `flushNowState` to `.notInFlushNow`, which should be able to be set unconditionally at the top of the function.

Reply:
good points, all defers and `flushNowState` gone. Changing the abstraction as you suggested instead of trying to keep the old (and busted) one alive was a much better call.